File: de.lmu.ifi.dbs.knowing.core/src/de/lmu/ifi/dbs/knowing/core/graph/Edge.java
package de.lmu.ifi.dbs.knowing.core.graph;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlRootElement;

@XmlRootElement
@XmlAccessorType(XmlAccessType.FIELD)
public class Edge {
	
	private transient Node source;
	private transient Node target;
	
	@XmlAttribute
	private String id;
	
	@XmlAttribute
	private String sourceId;
	
	@XmlAttribute
	private String targetId;
	
	@XmlAttribute
	private int weight;
	
	protected Edge() {
		this.id = String.valueOf(System.currentTimeMillis());
	}
	
	public Edge(String id,Node source, Node target) {
		this(id,source, target, 1);
	}
	
	public Edge(String id, String sourceId, String targetId) {
		this(id,sourceId, targetId, 1);
	}
	
	public Edge(String id, Node source, Node target, int weight) {
		this.id = id;
		this.source = source;
		this.target = target;
		this.weight = weight;
	}
		
	public Edge(String id, String sourceId, String targetId, int weight) {
		this.id = id;
		this.sourceId = sourceId;
		this.targetId = targetId;
		this.weight = weight;
	}

	public String getId() {
		return id;
	}
	
	public Node getSource() {
		return source;
	}

	public void setSource(Node source) {
		this.source = source;
	}

	public Node getTarget() {
		return target;
	}

	public void setTarget(Node target) {
		this.target = target;
	}

	public String getSourceId() {
		return sourceId;
	}

	public void setSourceId(String sourceId) {
		this.sourceId = sourceId;
		this.source = null;
	}

	public String getTargetId() {
		return targetId;
	}

	public void setTargetId(String targetId) {
		this.targetId = targetId;
		this.target = null;
	}

	public int getWeight() {
		return weight;
	}

	public void setWeight(int weight) {
		this.weight = weight;
	}
	
	@Override
	public Edge clone() {
//		if(source == null || target == null)
//			return new Edge(id, sourceId, targetId, weight);
//		else
//			return new Edge(id, source.clone(), target.clone());
		return new Edge(id, sourceId, targetId, weight);
	}

	@Override
	public String toString() {
		if(source == null && target == null)
			return sourceId + " to (" + weight + ")" + targetId;
		return source + " to (" + weight + ")" + target;
	}
	
	
}

File: de.lmu.ifi.dbs.knowing.core/src/de/lmu/ifi/dbs/knowing/core/graph/Edge.scala
package de.lmu.ifi.dbs.knowing.core.graph

import scala.annotation.target.field
import javax.xml.bind.annotation.{ XmlAccessorType, XmlAccessType, XmlAttribute }

@XmlAccessorType(XmlAccessType.FIELD)
class Edge(@(XmlAttribute @field) var id: String,
  @(XmlAttribute @field) var sourceId: String,
  @(XmlAttribute @field) var targetId: String,
  @(XmlAttribute @field) var weight: Int) extends Serializable {

  def this(id: String, sourceId: String, targetId: String) = this(id, sourceId, targetId, 1)

  def this() = this("", "", "", 0)

  def getPlainSourceId = sourceId.split(':')(0)
  def getPlainTargetId = targetId.split(':')(0)

  def setPlainSourceId(id: String) {
    hasSourcePort match {
      case true => sourceId = id + ":" + getSourcePort
      case false => sourceId = id
    }
  }

  def setPlainTargetId(id: String) {
    hasTargetPort match {
      case true => targetId = id + ":" + getTargetPort
      case false => targetId = id
    }
  }

  def getSourcePort: String = {
    if (hasSourcePort) return sourceId.split(':')(1)
    return ""
  }

  def getTargetPort: String = {
    if (hasTargetPort) return targetId.split(':')(1)
    return ""
  }

  def hasSourcePort: Boolean = sourceId contains (':')
  def hasTargetPort: Boolean = targetId contains (':')

  def setSourcePort(port: String) {
    port match {
      case null | "" => sourceId = getPlainSourceId
      case s: String => sourceId = getPlainSourceId + ":" + port
    }
  }
  def setTargetPort(port: String) {
    port match {
      case null | "" => targetId = getPlainTargetId
      case s: String => targetId = getPlainTargetId + ":" + port
    }
  }

  override def toString() = id + "[" + sourceId + " -> " + targetId + "]"

}